compiler/ccgcalls.nim

```diff
@@ -246,26 +255,72 @@ proc genArg(p: BProc, n: PNode, param: PSym; call: PNode): Rope =
       result = rdLoc(a)
   else:
     initLocExprSingleUse(p, n, a)
-    result = rdLoc(a)
+    result = rdLoc(withTmpIfNeeded(p, a, needsTmp))
 
-proc genArgNoParam(p: BProc, n: PNode): Rope =
+proc genArgNoParam(p: BProc, n: PNode, needsTmp = false): Rope =
   var a: TLoc
   if n.kind == nkStringToCString:
-    result = genArgStringToCString(p, n)
+    result = genArgStringToCString(p, n, needsTmp)
   else:
     initLocExprSingleUse(p, n, a)
-    result = rdLoc(a)
+    result = rdLoc(withTmpIfNeeded(p, a, needsTmp))
+
+import dfa
+from patterns import sameTrees
+
+proc potentialAlias(n: PNode, potentialWrites: seq[PNode]): bool =
+  for p in potentialWrites:
+    if sameTrees(p, n) or p.aliases(n) or n.aliases(p):
+      return true
+
+proc skipTrivialIndirections(n: PNode): PNode =
+  result = n
+  while true:
+    case result.kind
+    of {nkBracketExpr, nkDerefExpr, nkHiddenDeref,
+        nkAddr, nkHiddenAddr,
+        nkObjDownConv, nkObjUpConv}:
+      result = result[0]
+    of {nkHiddenStdConv, nkHiddenSubConv}:
+      result = result[1]
+    else: break
+
+proc genParams(p: BProc, ri: PNode, typ: PType): Rope =
+  # We must generate temporaries in cases like #14396
+  # to keep the strict Left-To-Right evaluation, this
+  # is a bit pessimistic currently
+  var needTmp = newSeq[bool](ri.len - 1)
+  var potentialWrites: seq[PNode]
+  for i in countdown(ri.len - 1, 1):
+    if ri[i].skipTrivialIndirections.kind == nkSym:
+      needTmp[i - 1] = potentialAlias(ri[i], potentialWrites)
+    else:
+      let cfg = constructCfg(ri[i])
+      var newPotentialWrites: seq[PNode]
+      # Simple linear pass, ignoring forks and gotos
```

Comment:
Er, why do you need the DFA then? There is only ``nkAsgn, nkFastAsgn`` and addr(x) that is potentially destructive. Easy enough to check for without the DFA.

Reply:
Done, though for some reason there are cases like `var f = 0; f += 1` where `f` is not an nkHiddenAddr, but an nkSym directly..